core/src/main/java/org/commonjava/maven/ext/core/impl/RelocationManipulator.java

```diff
@@ -230,37 +236,46 @@ private boolean updateDependencies( Project project, WildcardMap<ProjectVersionR
         return result;
     }
 
-    private void updateVersionString( Project project, ProjectVersionRef relocation, String property, Dependency dependency,
-                                      Node versionNode ) throws ManipulationException
+    private boolean updatePlugins( WildcardMap<ProjectVersionRef> pluginRelocations, final WildcardMap<ProjectVersionRef> dependencyRelocations, final Project project,
+                                   final Map<ProjectVersionRef, Plugin> pluginMap ) throws ManipulationException
     {
-        if ( property != null && property.contains( "$" ))
-        {
-            property = property.substring( 2, property.length() - 1 );
-            if ( StringUtils.countMatches( property, "${" ) > 1 )
-            {
-                throw new ManipulationException( "Relocations with multiple embedded version properties not supported" );
-            }
-            logger.debug( "Updating relocation for {} with property {} to new version {}",
-                          dependency == null ? versionNode : dependency, property, relocation.getVersionString() );
-            PropertiesUtils.updateProperties( session, project, true, property, relocation.getVersionString() );
-        }
-        else
+        // Handles plugins themselves
+        for ( int i = 0 ; i < pluginRelocations.size(); i++ )
```

Comment:
This loop (and the similar one below) doesn't look right to me. From what I can see, we're never accessing this value and we're not really iterating over this map, but I different one.

Reply:
Same pattern as used in dependencies so should be ok.

Reply:
I know the code has been there a long time, but it's not making sense to me. There's a double loop and the outer loop never appears to be used. The problem is `i` is never used. Is `relocations` ever used "sequentially" so that we're not looping `relocations.size()` extra times or something like this?

It has

```
        // If we do a single pass over the dependencies that will handle the relocations *but* it will not handle
        // where one relocation alters the dependency and a subsequent relocation alters it again. For instance, the
        // first might wildcard alter the groupId and the second, more specifically alters one with the artifactId
        for ( int i = 0 ; i < relocations.size(); i++ )
        {
            Iterator<ArtifactRef> it = dependencies.keySet().iterator();
            while ( it.hasNext() )
            {
                final ArtifactRef pvr = it.next();
                if ( relocations.containsKey( pvr.asProjectRef() ) )
                {
                    ProjectVersionRef relocation = relocations.get( pvr.asProjectRef() );

```



It's not as if `relocation = relocations.get( i )` as in a normal loop. Instead, `relocation =  relocations.get( pvr.asProjectRef() )` where `pvr` comes from the loop over `dependencies`, not `relocations`.

Reply:
Right but my point is you didn't change it in your original pr so why highlight it now? Currently I am not handling the iteration removal for plugins same as dependencies but reused the same pattern of code as I presumed its working.

Reply:
I've added a test for multiple plugin relocations and double checked the code.